---
title: "Visualization 4: karyoploteR"
output:
  html_document:
    df_print: paged
---


# karyoploteR

[karyoploteR](http://bioconductor.org/packages/karyoploteR/) is an 
R/Bioconductor package to plot data along the genome. It is data agnostic, 
meaning that it does not know anything about you data (it doesn't know bioogy!),
and thus is very flexible, because it makes no assumptions about your data. Oh,
and it's also organism agnostic. It can work with ANY genome, so very useful for
those working with non-model organisms!

It follows a philosophy similar to the base R graphics, where you start 
creating a plot and then add additional graphical elements on top of it.

You can get more information reading [the package vignette](http://bioconductor.org/packages/devel/bioc/vignettes/karyoploteR/inst/doc/karyoploteR.html),
reading the inline documentation with `?function_name` or you can go to the
karyoploteR tutorial page at [https://bernatgel.github.io/karyoploter_tutorial/](https://bernatgel.github.io/karyoploter_tutorial/), where you'll also find examples of somewhat complex plots. If 
everything else fail, ask a question on [biostars](https://www.biostars.org/) 
with the tag "karyoploter" or send me an email to bgel@igtp.cat!


To install the package use BiocManager
```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("karyoploteR")
```

and once it's installed, load it with `library()`

```{r}
library(karyoploteR)
```



## Let's get started

To create our first karyoplot we'll call the function `plotKaryotype`.

```{r}
kp <- plotKaryotype()
```

  By default it will create a karyoplot of the Human genome version hg19 (GRCh37). To 
  use another genome we can specify its name, for example "mm10" for mouse
  
```{r}
kp <- plotKaryotype(genome="mm10")
```
  
or for drosophila, "dm6"

```{r}
available.genomes()
library(BSgenome)
kp <- plotKaryotype(genome="dm6")
```

For a few organisms karyoploteR has the chromosomes information. If it
has no information on about a genome, it will try to load the corresponding
BSgenome object and if that is not possible, it will fail.

For such cases, we can use a custom genome, a GRanges object with one
range per chromosome.

```{r}
custom.genome <- toGRanges(c("A:1-1000", "B:1-800"))
kp <- plotKaryotype(genome=custom.genome)
kpAddBaseNumbers(kp, tick.dist = 100)
```

It is also possible to plot only a few chromosomes using the... "chromosomes"
parameter.

```{r}
kp <- plotKaryotype(chromosomes=c("chr3", "chr5", "chr17"))
```

## Add some data!

Once we have a karyoplot we can add data into it with any of the the available 
plotting functions. For example, one point

```{r}
kp <- plotKaryotype(chromosomes=c("chr1", "chr2"))
kpPoints(kp, chr="chr1", x=10e6, y=0.5, col="red")
```

or many

```{r}
tiles <- unlist(tileGenome(kp$chromosome.lengths, tilewidth = 1e5))
y <- rnorm(length(tiles), mean = 0.5, sd=0.1)
y[1400:1600] <- y[1400:1600]+0.2
kp <- plotKaryotype(chromosomes=c("chr1", "chr2"))
kpPoints(kp, data = tiles, y=y, col="blue")
```

And we can control the appearance of the graphics with the same standard graphics
parameters we have seen before.

```{r}
kp <- plotKaryotype(chromosomes=c("chr1"))
kpPoints(kp, chr="chr1", x=1:20*10e6, y=1:20*1/20, col=rainbow(20), cex=1:20/3, pch=1:20)
```

We can use `kpLines`, `kpArrows`, `kpText`, `kpRect`... to plot simple graphical primitives on the genome.

## And what about real data?

karyoploteR has a second group of plotting function that start with "kpPlot".
This functions take a sligthly more complex data, usually perform some kind 
of small computation and plot. 

We'll see a few examples

### Plotting genes

```{r}
  library(TxDb.Hsapiens.UCSC.hg19.knownGene)
  library(org.Hs.eg.db)
  all.genes <- filterChromosomes(genes(TxDb.Hsapiens.UCSC.hg19.knownGene))
  all.genes
```

and select randomly 20 of them as our reeeeally favourite genes

```{r}
  my.genes <- sample(all.genes, size = 20)
  my.genes
```

Now get the symbol of these genes using from the Entrez GeneId identifier

```{r}
  symbols <- select(org.Hs.eg.db, keys=my.genes$gene_id, keytype="ENTREZID", columns="SYMBOL")
  my.genes$symbol <- symbols[,2]
  my.genes
```

and plot them using `kpPlotMarkers`

```{r}
kp <- plotKaryotype()
kpPlotMarkers(kp, data=my.genes, labels = my.genes$symbol, text.orientation = "horizontal", cex=0.5 , r1=0.6)
```


### Plot the density of reads in a BAM file

Given a BAM file, how are the reads distributed accross the genome?

```{r}
kp <- plotKaryotype()
kpPlotBAMDensity(kp, data="datasets/ChIP-seq_sample/NL1_h3K27ac_sample.rmdup.bam", col="dodgerblue")
```

```{r}
kp <- plotKaryotype(chromosomes="chr10")
kpAddBaseNumbers(kp)
kpPlotBAMDensity(kp, data="datasets/ChIP-seq_sample/NL1_h3K27ac_sample.rmdup.bam", col="dodgerblue")
```


```{r}
kp <- plotKaryotype(zoom="chr10:40e6-45e6")
kpAddBaseNumbers(kp, tick.dist = 5e5)
kpPlotBAMDensity(kp, data="datasets/ChIP-seq_sample/NL1_h3K27ac_sample.rmdup.bam", col="dodgerblue", window.size = 10000)
```



### Plot Regions on the genome

As an example, plot the peaks you found yesterday

```{r}
peaks <- toGRanges("datasets/ChIP-seq_sample/peaks/NL1_h3K27ac_peaks.broadPeak")

kp <- plotKaryotype(chromosomes = "chr1")
kpPlotRegions(kp, data=peaks)
```













